### PR TITLE
FIX #200: Support Mongodb Persistence for WF

### LINKED ISSUE #200: Support Mognodb Persistence for Workflows
Since the workflow state is defined as a JSON object, I think a Mongodb is a suitable option to be used a persistence

### PR DESCRIPTION
**Pull request by manarhusrieh**
I was using the library workflow part and came across the case where I needed to use Mongodb as a persistence for states. 

High level changes: 
* Add new package for @node-ts/bus-mongodb.
* Add mongodb configuration.
* Add mongodb persistence.
* Add integration tests for mongodb.

@adenhertog  I have copied this implementation from bus-postgres package that exists in the project.

### CODE REVIEW

File: packages/bus-mongodb/README.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
+# @node-ts/bus-postgres
```
Comment (adenhertog) [inline packages/bus-mongodb/README.md:1]:
```suggestion
# @node-ts/bus-mongodb
```

File: packages/bus-mongodb/README.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+## Installation
+
+Install all packages and their dependencies
+
+```bash
+npm install @node-ts/bus-mongodb
+```
+
+Configure a new Postgres persistence and register it with `Bus`:
```
Comment (adenhertog) [inline packages/bus-mongodb/README.md:17]:
```suggestion
Configure a new MongoDB persistence and register it with `Bus`:
```

File: packages/bus-mongodb/README.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+```typescript
+import { Bus } from '@node-ts/bus-core'
+import { MongodbPersistence, MongodbConfiguration } from '@node-ts/bus-postgres'
```
Comment (adenhertog) [inline packages/bus-mongodb/README.md:21]:
```suggestion
import { MongodbPersistence, MongodbConfiguration } from '@node-ts/bus-mongodb'
```

File: packages/bus-mongodb/src/mongodb-persistence.integration.ts [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+  MessageWorkflowMapping,
+  Logger,
+  BusInstance,
+} from "@node-ts/bus-core";
+import { MessageAttributes } from "@node-ts/bus-messages";
+import { MongodbPersistence } from "./mongodb-persistence";
+import { MongodbConfiguration } from "./mongodb-configuration";
+import { Mock } from "typemoq";
+import { TestWorkflowState, TestCommand, TestWorkflow } from "../test";
+import { Pool } from "pg";
```
Comment (adenhertog) [inline packages/bus-mongodb/src/mongodb-persistence.integration.ts:13]:
```suggestion
```

File: packages/bus-mongodb/src/mongodb-persistence.ts [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+    );
+    this.client;
+  }
+
+  async initialize(): Promise<void> {
+    this.logger.info("Initializing mongodb persistence...");
+    if (this.client) {
+      await this.client?.connect();
+      this.database = this.client?.db(this.configuration.databaseName);
+    }
```
Comment (adenhertog) [inline packages/bus-mongodb/src/mongodb-persistence.ts:43]: should this throw an error if `this.database` is never assigned? `initialize()` should only really complete successfully if the persistence is ready to go

File: packages/bus-mongodb/src/mongodb-persistence.ts [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+    this.logger.info("Mongodb persistence initialized");
+  }
+
+  async dispose(): Promise<void> {
+    this.logger.info("Disposing Mongodb persistence...");
+    if (this.client) {
+      this.client.close();
```
Comment (adenhertog) [inline packages/bus-mongodb/src/mongodb-persistence.ts:50]:
```suggestion
      await this.client.close();
```

File: packages/bus-mongodb/src/mongodb-persistence.ts [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+    if (!collectionExists) {
+      await this.database?.createCollection(collectionName);
+    }
+  }
+
+  private async ensureIndexesExist(
+    collectionName: string,
+    messageWorkflowMappings: MessageWorkflowMapping<Message, WorkflowState>[]
+  ): Promise<void> {
+    const collection = this.database?.collection(collectionName);
```
Comment (adenhertog) [inline packages/bus-mongodb/src/mongodb-persistence.ts:161]: this actually runs before `initialize()`, so `this.database` is always undefined and none of these indexes are actually created. I'd recommend:

1. don't use optional chaining (`this.database?.`) but instead the non-null assertion throughout this method eg:
```suggestion
    const collection = this.database!.collection(collectionName);
```

**Comment by adenhertog** [comment]
hey @manarhusrieh thanks for the PR - looking forward to getting this merged.

I've checked it out locally to have a bit of a play with it and it looks really good. Just left a few comments for your consideration, so let me know when you'd like me to have another look :)

**Comment by manarhusrieh** [comment]
@adenhertog thanks for looking at the PR and I will fix these comments. Regarding commits, do you want me to amend the commit and push again? Or shall I just push a new commit and you will squash them during merge?

**Comment by adenhertog** [comment]
Hey @manarhusrieh, whatever you prefer. I'll end up squash committing, so feel free to push as many commits to this PR as you need

File: packages/bus-mongodb/src/mongodb-persistence.ts (hunk 7 above)
Comment (manarhusrieh) [reply packages/bus-mongodb/src/mongodb-persistence.ts:161]: Can you tell me in which case the initialize method runs after initializeWorkflow()? @adenhertog

Comment (adenhertog) [reply packages/bus-mongodb/src/mongodb-persistence.ts:161]: `initializeWorkflow()` always runs before `initialize()`. You should be able to see this at `workflow-registry.ts:112` and `workflow-registry.ts:120`

File: packages/bus-mongodb/README.md (hunk 2 above)
Comment (manarhusrieh) [reply packages/bus-mongodb/README.md]: Fixed

File: packages/bus-mongodb/README.md (hunk 3 above)
Comment (manarhusrieh) [reply packages/bus-mongodb/README.md]: Fixed

File: packages/bus-mongodb/src/mongodb-persistence.ts (hunk 5 above)
Comment (manarhusrieh) [reply packages/bus-mongodb/src/mongodb-persistence.ts]: Fixed the solution and applied a new way to handle it plz check it @adenhertog

File: packages/bus-mongodb/src/mongodb-persistence.ts (hunk 6 above)
Comment (manarhusrieh) [reply packages/bus-mongodb/src/mongodb-persistence.ts]: Fixed

File: packages/bus-mongodb/src/mongodb-persistence.ts (hunk 7 above)
Comment (manarhusrieh) [reply packages/bus-mongodb/src/mongodb-persistence.ts]: Fixed plz check the solution @adenhertog

**Comment by manarhusrieh** [comment]
Hey @adenhertog, I have pushed a new commit to handle the changes you have requested. Please review again.

**Comment by adenhertog** [comment]
looks great! thanks @manarhusrieh I'll roll a new version with this in it

**Pull request merged**